docs/csharp/tutorials/exploration/csharp-7.yml

```diff
@@ -0,0 +1,263 @@
+
+    Test it with something like the following code:
+
+    [!code-csharp[TestWhenClause](~/samples/snippets/csharp/new-in-7/patternmatch.cs#TestWhenClause "Test the when clause")]
+
+    The new syntax for pattern matching expressions makes it easier to create dispatch algorithms based on an object's type, or other properties, using a clear and concise syntax. Pattern matching expressions enable these constructs on data types that are unrelated by inheritance.
+
+    You can learn more about pattern matching in the article dedicated to [pattern matching in C#](../../pattern-matching.md).
+
+- title: Optimize memory storage using ref locals and returns
+  durationInMinutes: 2
+  content: |
+    This feature enables algorithms that use and return references to variables defined elsewhere. One example is with large matrices and finding a single location with certain characteristics. One method would return the two indices a single location in the matrix:
+
+    [!code-csharp[FindReturningIndices](~/samples/snippets/csharp/new-in-7/MatrixSearch.cs#EverythingByValue "Find returning indices")]
+
+    This method returns the indices to the item in the matrix. That leads callers to write code that uses those indices to dereference the matrix and modify a single element:
+
+    [!code-csharp[TestReturningIndices](~/samples/snippets/csharp/new-in-7/MatrixSearch.cs#TestByValue "Update Item From Indices")]
+
+    You'd rather write a method that returns a *reference* to the element of the matrix that you want to change.
+
+    Let's walk through a series of changes to demonstrate the ref local feature and show how to create a method that returns a reference to internal storage. Along the way, you'll learn the rules of the ref return and ref local feature that protect you from accidentally misusing it.
+
+    Start by modifying the `Find` method declaration so that it returns a `ref int` instead of a tuple.
+    
+    ```csharp
+    ref int Find(int[,] matrix, Func<int, bool> predicate)
+    ```
+    
+    Modify the return statement to return the item at the correct indices:
+
+    ```csharp
+    return matrix[i,j];
+    ```
+    
+    Change the final return to throw an exception instead:
+
+    ```csharp
+    throw new InvalidOperationException("Not found");
+    ```
+
+    Note that this won't compile. The method declaration indicates a `ref` return, but the return statement specifies a value return. You must add the `ref` keyword to each return statement. That indicates return by reference, and helps developers reading the code later remember that the method returns by reference:
+
+    ```csharp
+    return ref matrix[i,j];
+    ```
+
+    Now that the method returns a reference to the integer value in the matrix, you need to modify where it's called.  The `var` declaration means that `valItem` is now an `int` rather than a tuple. Change the calling code to following:
```

Comment:
```suggestion
    Now that the method returns a reference to the integer value in the matrix, you need to modify where it's called. The `var` declaration means that `valItem` is now an `int` rather than a tuple. Change the calling code to following:
```